File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/KvtxScriptUtils.swift
// Copyright 2026 Polar Electro Oy. All rights reserved.

import Foundation

private let TAG = "KvtxScriptUtils"

/// Generic KVTXScript builder and scanner.
internal enum KvtxScriptUtils {

    static let CMD_WRITE_BYTES:    UInt8 = 0x00
    static let CMD_APPEND_BYTES:   UInt8 = 0x01
    static let CMD_REMOVE:         UInt8 = 0x02
    static let CMD_COPY:           UInt8 = 0x03
    static let CMD_MOVE:           UInt8 = 0x04
    static let CMD_COMMIT:         UInt8 = 0x05
    static let CMD_WRITE_BYTES_EX: UInt8 = 0x06
    static let CMD_APPEND_BYTES_EX:UInt8 = 0x07
    static let CMD_REMOVE_EX:      UInt8 = 0x08

    /// Build a KVTXScript that writes `data` to `kvKey` and commits.
    ///
    /// Structure: WRITE_BYTES(key, data) + COMMIT
    static func buildWriteAndCommit(kvKey: UInt32, data: [UInt8]) -> [UInt8] {
        var script = [UInt8]()
        script.append(CMD_WRITE_BYTES)
        script.append(contentsOf: u32Le(kvKey))
        script.append(contentsOf: u32Le(UInt32(data.count)))
        script.append(contentsOf: data)
        script.append(CMD_COMMIT)
        return script
    }

    /// Scan a full KVTXScript binary and extract the raw value bytes stored under `kvKey`.
    ///
    /// Returns `nil` if the key is not present (or was removed) in the script.
    static func extractValueForKey(script: [UInt8], kvKey: UInt32) -> [UInt8]? {
        var pos = 0
        var result: [UInt8]? = nil

        func readUInt32() -> UInt32? {
            guard pos + 4 <= script.count else { return nil }
            let v = UInt32(script[pos]) |
                    (UInt32(script[pos+1]) << 8) |
                    (UInt32(script[pos+2]) << 16) |
                    (UInt32(script[pos+3]) << 24)
            pos += 4
            return v
        }

        func readBytes(_ n: Int) -> [UInt8]? {
            guard n >= 0, pos + n <= script.count else { return nil }
            let bytes = Array(script[pos..<(pos+n)])
            pos += n
            return bytes
        }

        while pos < script.count {
            let startPos = pos
            let cmd = script[pos]; pos += 1

            switch cmd {
            case CMD_WRITE_BYTES, CMD_APPEND_BYTES:
                guard let key = readUInt32(), let length = readUInt32(), let data = readBytes(Int(length)) else {
                    BleLogger.trace("\(TAG): extractValueForKey: parse error at \(startPos) — stopping")
                    return result
                }
                if key == kvKey {
                    result = (cmd == CMD_WRITE_BYTES) ? data : (result ?? []) + data
                }

            case CMD_REMOVE:
                guard let key = readUInt32() else { return result }
                if key == kvKey { result = nil }

            case CMD_COPY, CMD_MOVE:
                guard readUInt32() != nil, readUInt32() != nil else { return result }

            case CMD_COMMIT:
                break // no payload

            case CMD_WRITE_BYTES_EX, CMD_APPEND_BYTES_EX:
                guard let key = readUInt32() else { return result }
                guard pos < script.count else { return result }
                let idxLen = Int(script[pos]); pos += 1
                guard let _ = readBytes(idxLen), let length = readUInt32(), let data = readBytes(Int(length)) else { return result }
                if key == kvKey && idxLen == 0 {
                    result = (cmd == CMD_WRITE_BYTES_EX) ? data : (result ?? []) + data
                }

            case CMD_REMOVE_EX:
                guard let key = readUInt32() else { return result }
                guard pos < script.count else { return result }
                let idxLen = Int(script[pos]); pos += 1
                guard readBytes(idxLen) != nil else { return result }
                if key == kvKey && idxLen == 0 { result = nil }

            default:
                BleLogger.trace("\(TAG): extractValueForKey: unknown command 0x\(String(cmd, radix: 16)) @ \(startPos) — stopping")
                return result
            }
        }
        return result
    }

    // MARK: - Helpers

    static func u32Le(_ value: UInt32) -> [UInt8] {
        return [
            UInt8(value & 0xFF),
            UInt8((value >> 8) & 0xFF),
            UInt8((value >> 16) & 0xFF),
            UInt8((value >> 24) & 0xFF)
        ]
    }
}

File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/KvtxScriptUtils.kt
// Copyright © 2026 Polar Electro Oy. All rights reserved.
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

private const val TAG = "KvtxScriptUtils"

/**
 * Generic KVTXScript builder and scanner.
 */
internal object KvtxScriptUtils {

    const val CMD_WRITE_BYTES: Byte = 0x00
    const val CMD_APPEND_BYTES: Byte = 0x01
    const val CMD_REMOVE: Byte = 0x02
    const val CMD_COPY: Byte = 0x03
    const val CMD_MOVE: Byte = 0x04
    const val CMD_COMMIT: Byte = 0x05
    const val CMD_WRITE_BYTES_EX: Byte = 0x06
    const val CMD_APPEND_BYTES_EX: Byte = 0x07
    const val CMD_REMOVE_EX: Byte = 0x08

    /**
     * Build a KVTXScript that writes [data] to [kvKey] and commits.
     *
     * Structure: WRITE_BYTES(key, data) + COMMIT
     */
    fun buildWriteAndCommit(kvKey: Int, data: ByteArray): ByteArray =
        ByteArrayOutputStream().apply {
            write(byteArrayOf(CMD_WRITE_BYTES))
            write(u32Le(kvKey))
            write(u32Le(data.size))
            write(data)
            write(byteArrayOf(CMD_COMMIT))
        }.toByteArray()

    /**
     * Scan a full KVTXScript binary and extract the raw value bytes stored under [kvKey].
     *
     * Returns `null` if the key is not present (or was removed) in the script.
     */
    fun extractValueForKey(script: ByteArray, kvKey: Int): ByteArray? {
        val buf = ByteBuffer.wrap(script).order(ByteOrder.LITTLE_ENDIAN)
        val targetKey = kvKey.toUInt().toLong()
        var result: ByteArray? = null

        while (buf.hasRemaining()) {
            val pos = buf.position()
            when (val cmdByte = buf.get().toInt() and 0xFF) {
                CMD_WRITE_BYTES.toInt(), CMD_APPEND_BYTES.toInt() -> {
                    val key = buf.int; val length = buf.int
                    val data = ByteArray(length); buf.get(data)
                    if (key.toUInt().toLong() == targetKey) {
                        result = if (cmdByte == CMD_WRITE_BYTES.toInt()) data else (result ?: byteArrayOf()) + data
                    }
                }
                CMD_REMOVE.toInt() -> {
                    val key = buf.int
                    if (key.toUInt().toLong() == targetKey) result = null
                }
                CMD_COPY.toInt(), CMD_MOVE.toInt() -> { buf.int; buf.int }
                CMD_COMMIT.toInt() -> { }
                CMD_WRITE_BYTES_EX.toInt(), CMD_APPEND_BYTES_EX.toInt() -> {
                    val key = buf.int
                    val idxLen = buf.get().toInt() and 0xFF
                    val idxBytes = ByteArray(idxLen); buf.get(idxBytes)
                    val length = buf.int
                    val data = ByteArray(length); buf.get(data)
                    if (key.toUInt().toLong() == targetKey && idxLen == 0) {
                        result = if (cmdByte == CMD_WRITE_BYTES_EX.toInt()) data else (result ?: byteArrayOf()) + data
                    }
                }
                CMD_REMOVE_EX.toInt() -> {
                    val key = buf.int
                    val idxLen = buf.get().toInt() and 0xFF
                    val idxBytes = ByteArray(idxLen); buf.get(idxBytes)
                    if (key.toUInt().toLong() == targetKey && idxLen == 0) result = null
                }
                else -> {
                    BleLogger.w(TAG, "extractValueForKey: unknown command 0x${cmdByte.toString(16)} @ $pos — stopping")
                    break
                }
            }
        }
        return result
    }

    fun u32Le(value: Int): ByteArray =
        ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array()
}

